presto-spi/src/main/java/com/facebook/presto/spi/block/MapBlockBuilder.java

```diff
@@ -504,6 +504,19 @@ private static int getHashPosition(Block keyBlock, int position, MethodHandle ke
             throw new RuntimeException(throwable);
         }
 
-        return (int) Math.floorMod(hashCode, hashTableSize);
+        return getReducedHashPosition(hashCode, hashTableSize);
+    }
+
+    // Adapted from the fastrange algorithm: https://github.com/lemire/fastrange. This function reduces the hashcode
+    // to [0, hashTableSize) uniformly. It first gets a 32 bit hashcode by calling Long.hashCode(), then reduce this 32
+    // bit hash code to [0, hashTableSize) by multiplying it by hashTableSize then divided by 2^32. By multiplying by
+    // hashTableSize, we take 32 bit hashcode in the range [0, 2^32) and map them to multiples of hashTableSize in
+    // [0, hashTableSize * 2^32). By dividing by 2^32, we map all multiples of hashTableSize in [0, 2^32) to 0, all
+    // multiples of hashTableSize in [2^32, 2 * 2^32) to one, and so forth. In each interval of 2^32 size, the number of
+    // reduced values must be 2^32/hashTableSize, therefore the distribution is uniform if the hashcode distribution is
+    // uniform.
+    static int getReducedHashPosition(long hashcode, int hashTableSize)
+    {
+        return (int) (((Long.hashCode(hashcode) & 0xffff_ffffL) * (long) hashTableSize) >> 32);
```

Comment:
IIRC originally the constant is `0x7fff_ffffL` to project the value into [0, 2^31)?

If we don't need to project to [0, 2^31), do we still need this bitwise and step?

Reply:
If and with 0x7fff_ffffL, then we need to shift 31 bits. If and with 0xffff_ffffL we need to shift 32 bits. The reduced range will both be [0, hashtableSize). I wrote it as 0x7fff_ffff before because there was a bug: I originally used 0xffff_ffff, but missed the 'L' after the 0xffff_ffff, such that (long)(Long.hashCode(hashcode)) & 0xffff_ffff could still be 64 bits something like 0xffff_ffff_ff08_ac50 while what we wanted was something like 0x0000_0000_ff08_ac50. This is because 0xffff_ffff is -1 in 32bit, and it was actually 0xffff_ffff_ffff_ffffL before anding with the first part. But if it's 0x7fff_ffff then it doesn't require the 'L'. At the time I didn't know why it was the case so I used 0x7fff_ffff. Then I realized it was missing the 'L'. With 0xffff_ffffL the high 32 bits are all zeros and we will get what we want.

Reply:
@yingsu00 : Ah I see, so you want to extend it to 64bit, but don't want to keep all high 32 bit to be 0 (instead of extending the signed bit.

Reply:
@wenleix yes. If we can make hashcode max 2^32-1 then the multiplication won't overflow because hashTableSize max is 2^31-1.